core/src/main/scala/org/locationtech/rasterframes/expressions/localops/Resample.scala

```diff
@@ -22,53 +22,168 @@
-import org.locationtech.rasterframes.expressions.DynamicExtractors.tileExtractor
+import org.apache.spark.sql.rf.TileUDT
+import org.apache.spark.sql.types.{DataType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+import org.locationtech.rasterframes.encoders.CatalystSerializer._
+import org.locationtech.rasterframes.expressions.{fpTile, row}
+import org.locationtech.rasterframes.expressions.DynamicExtractors._
+
+
+abstract class ResampleBase(left: Expression, right: Expression, method: Expression)
+  extends TernaryExpression
+  with CodegenFallback with Serializable {
 
-@ExpressionDescription(
-  usage = "_FUNC_(tile, factor) - Resample tile to different size based on scalar factor or tile whose dimension to match. Scalar less than one will downsample tile; greater than one will upsample. Uses nearest-neighbor value.",
-  arguments = """
-  Arguments:
-    * tile - tile
-    * rhs  - scalar or tile to match dimension""",
-  examples = """
-  Examples:
-    > SELECT _FUNC_(tile, 2.0);
-       ...
-    > SELECT _FUNC_(tile1, tile2);
-       ..."""
-)
-case class Resample(left: Expression, right: Expression) extends BinaryLocalRasterOp
-  with CodegenFallback {
   override val nodeName: String = "rf_resample"
-  override protected def op(left: Tile, right: Tile): Tile = left.resample(right.cols, right.rows, NearestNeighbor)
-  override protected def op(left: Tile, right: Double): Tile = left.resample((left.cols * right).toInt,
-                                                                             (left.rows * right).toInt, NearestNeighbor)
-  override protected def op(left: Tile, right: Int): Tile = op(left, right.toDouble)
+  override def dataType: DataType = left.dataType
+  override def children: Seq[Expression] = Seq(left, right, method)
+
+  def targetFloatIfNeeded(t: Tile, method: ResampleMethod): Tile =
+     method match {
+      case NearestNeighbor | Mode | RMax | RMin | Sum  ⇒ t
+      case _ ⇒ fpTile(t)
+     }
+
+  def stringToMethod(methodName: String): ResampleMethod =
+    methodName.toLowerCase().trim().replaceAll("_", "") match {
+      case "nearestneighbor" | "nearest" ⇒ NearestNeighbor
+      case "bilinear" ⇒ Bilinear
+      case "cubicconvolution" ⇒ CubicConvolution
+      case "cubicspline" ⇒ CubicSpline
+      case "lanczos" | "lanzos" ⇒ Lanczos
+      // aggregates
+      case "average" ⇒ Average
+      case "mode" ⇒ Mode
+      case "median" ⇒ Median
+      case "max" ⇒ RMax
+      case "min" ⇒ RMin
+      case "sum" ⇒ Sum
+    }
+
+   // These methods define the core algorithms to be used.
+  def op(left: Tile, right: Tile, method: String): Tile = {
+    val m = stringToMethod(method)
+    targetFloatIfNeeded(left, m)
+      .resample(right.cols, right.rows, m)
+  }
+
+  def op(left: Tile, right: Double, method: String): Tile = {
+    val m = stringToMethod(method)
+    targetFloatIfNeeded(left, m)
+      .resample((left.cols * right).toInt, (left.rows * right).toInt, m)
```

Comment:
For documentation sake, can you introduce an intermediate variable for `left.cols * right` and `left.rows * right`? Something like `newRows` and `newCols` (if I'm capturing the intent properly).